// stepzen/lib/commands/init.js
"use strict";
// Copyright (c) 2020,2021,2022, StepZen, Inc.
Object.defineProperty(exports, "__esModule", { value: true });
const command_1 = require("@oclif/command");
const workspace_1 = require("../shared/workspace");
const zen_command_1 = require("../shared/zen-command");
class Init extends zen_command_1.default {
    async run() {
        const { args, flags } = this.parse(Init);
        const created = await (0, workspace_1.initWorkspace)({
            directory: args.directory,
            endpoint: flags.endpoint,
            yes: flags.yes,
        });
        this.log(`Created a StepZen workspace in ${created.directory}`);
        return created;
    }
}
exports.default = Init;
Init.description = 'Initialize a StepZen workspace in the current directory';
Init.hidden = true;
Init.flags = Object.assign(Object.assign({}, zen_command_1.default.flags), { endpoint: command_1.flags.string({ hidden: true }), help: command_1.flags.help({ char: 'h' }), yes: command_1.flags.boolean({ default: false, hidden: true }) });
Init.args = [
    {
        hidden: true,
        name: 'directory',
    },
];
