// stepzen/lib/commands/start.js
"use strict";
// Copyright (c) 2020,2021,2022, StepZen, Inc.
Object.defineProperty(exports, "__esModule", { value: true });
const chalk = require("chalk");
const chokidar = require("chokidar");
const debug = require("debug");
const detect = require("detect-port");
const dotenv = require("dotenv");
const path = require("path");
const prettyMilliseconds = require("pretty-ms");
const command_1 = require("@oclif/command");
const core_1 = require("@oclif/core");
const errors_1 = require("@oclif/errors");
const lodash_1 = require("lodash");
const deploy_1 = require("../commands/deploy");
const validate_1 = require("../commands/validate");
const constants_1 = require("../shared/constants");
const zen_command_1 = require("../shared/zen-command");
const utils_1 = require("../shared/utils");
const dashboard = require('@stepzen/graphiql-proxy');
const { version } = require('../../package.json');
class Start extends zen_command_1.default {
    async run() {
        const { flags } = this.parse(Start);
        const { configuration } = await this.ensureStepZenAccount();
        // Get or create a StepZen workspace (possibly interactive)
        const workspace = await this.ensureStepZenWorkspace({
            directory: flags.dir,
            endpoint: flags.endpoint,
        });
        // If the user has overridden the endpoint, apply the override
        if (flags.endpoint) {
            workspace.endpoint = flags.endpoint;
        }
        if (!flags['no-dashboard']) {
            // Check the port is available
            const port = await detect(flags.port);
            if (flags.port !== port) {
                throw new errors_1.CLIError(`Could not start - port ${flags.port} is already in use`);
            }
        }
        const printEndpointDetailsOnce = (0, lodash_1.once)(() => {
            this.log((0, utils_1.formatEndpointInfo)({
                account: configuration.account,
                endpoint: workspace.endpoint,
                proxyport: flags['no-dashboard'] ? undefined : flags.port,
            }));
            this.log();
        });
        // This is the file watcher
        if (!flags['no-watcher']) {
            const redeploy = (0, lodash_1.debounce)((path) => {
                debug('stepzen:start')(`chokidar decected change to ${path}`);
                console.log(`File changed: ${chalk.blue((0, utils_1.workspaceRelative)(path, workspace.directory))}`);
                this.deployOnce({ workspace, flags }).then(didDeploy => {
                    didDeploy && printEndpointDetailsOnce();
                });
            }, 500);
            chokidar
                .watch(['**/*.graphql', 'config.yaml'], {
                cwd: workspace.schema,
                ignored: 'node_modules/**',
                ignoreInitial: true,
            })
                .on('change', redeploy)
                .on('add', redeploy)
                .on('unlink', redeploy);
            debug('stepzen:start')(`chokidar started watching ${workspace.schema}`);
        }
        // Start!
        core_1.CliUx.ux.action.start('Starting...');
        let didDeploy = false;
        // Unless explicitly disabled, auto-init
        if (!flags['no-init']) {
            didDeploy = await this.deployOnce({ workspace, flags });
        }
        // Create the dashboard
        if (!flags['no-dashboard']) {
            debug('stepzen:start')(`starting the dashboard server on ${flags.port}`);
            const ui = await dashboard({
                account: configuration.account,
                adminkey: configuration.adminkey,
                apikey: configuration.apikey,
                cli: {
                    version,
                },
                domain: constants_1.STEPZEN_DOMAIN,
                predicates: {
                    available: true,
                    enabled: false,
                    onTogglePredicates: async () => null,
                },
                port: flags.port,
                workspace,
            });
            await ui.start();
            debug('stepzen:start')(`dashboard server on ${flags.port} started`);
        }
        // Done
        core_1.CliUx.ux.action.stop();
        if (!flags['no-console']) {
            if (didDeploy) {
                printEndpointDetailsOnce();
            }
            if (!flags['no-watcher']) {
                console.log(`Watching ${chalk.blue((0, utils_1.homeRelative)(workspace.schema))} for changes...`);
            }
        }
    }
    async deployOnce({ workspace, flags, }) {
        dotenv.config({ path: path.join(workspace.directory, '.env') });
        if (!flags['no-validate']) {
            const validateCmd = `stepzen validate ${workspace.schema}`;
            try {
                debug('stepzen:start')(`starting $(${validateCmd})`);
                await validate_1.default.run([workspace.schema]);
                debug('stepzen:start')(`$(${validateCmd}) completed successfully`);
            }
            catch (error) {
                debug('stepzen:start')(`$(${validateCmd}) failed with`, error);
                console.log();
                console.log(chalk.red('Your local schema has the following GraphQL errors:'));
                console.log();
                console.log(error.message);
                return false;
            }
        }
        const logError = (error) => {
            try {
                const zenctlError = JSON.parse(error.message);
                const msg = zenctlError[0].message || zenctlError[0] || zenctlError;
                console.log('');
                console.log(chalk.red(msg));
            }
            catch (_a) {
                console.log('');
                console.log(chalk.red(error));
            }
        };
        const deployStart = new Date().getTime();
        core_1.CliUx.ux.action.start(`Deploying ${chalk.yellow(workspace.endpoint)} to StepZen`);
        const deployCmdArgs = flags.endpoint ? [flags.endpoint] : [];
        deployCmdArgs.push('--dir', workspace.directory, '--silent');
        const deployCmd = `stepzen deploy '${deployCmdArgs.join("' '")}'`;
        try {
            debug('stepzen:start')(`starting $(${deployCmd})`);
            await deploy_1.default.run(deployCmdArgs);
            debug('stepzen:start')(`$(${deployCmd}) completed successfully`);
        }
        catch (error) {
            debug('stepzen:start')(`$(${deployCmd}) failed with`, error);
            core_1.CliUx.ux.action.stop(chalk.red('fail'));
            logError(error);
            return false;
        }
        const deployEnd = new Date().getTime();
        const deployTime = deployEnd - deployStart;
        core_1.CliUx.ux.action.stop(`${chalk.grey('done in')} ${prettyMilliseconds(deployTime)} 🚀`);
        return true;
    }
}
exports.default = Start;
Start.description = 'deploy your schema to StepZen, and then redeploy on every change';
Start.flags = Object.assign(Object.assign({}, zen_command_1.default.flags), { dir: command_1.flags.string({ description: 'working directory' }), endpoint: command_1.flags.string({ description: 'Override workspace endpoint' }), help: command_1.flags.help({ char: 'h' }), 'no-console': command_1.flags.boolean({ hidden: true }), 'no-dashboard': command_1.flags.boolean({ hidden: true }), 'no-init': command_1.flags.boolean({ hidden: true }), 'no-validate': command_1.flags.boolean({ hidden: true }), 'no-watcher': command_1.flags.boolean({ hidden: true }), port: command_1.flags.integer({ default: 5001, env: 'PORT' }) });
Start.args = [];
